modules/mod-deep-learning/DeepModel.cpp

```diff
@@ -0,0 +1,199 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   DeepModel.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+
+#include "DeepModel.h"
+
+#include <torch/script.h>
+#include <torch/torch.h>
+
+#include "FileNames.h"
+#include "WaveTrack.h"
+#include "WaveClip.h"
+
+#include "CodeConversions.h"
+
+// DeepModel Implementation
+
+FilePath DeepModel::DLModelsDir()
+{
+   return FileNames::MkDir( wxFileName( FileNames::DataDir(), wxT("deeplearning-models") ).GetFullPath() );
+}
+
+FilePath DeepModel::BuiltInModulesDir()
+{
+   return FileNames::MkDir( wxFileName( FileNames::BaseDir(), wxT("deeplearning-models") ).GetFullPath() );
+}
+
+void DeepModel::LoadResampler()
+{
+   // load the resampler module
+   std::string resamplerPath = audacity::ToUTF8(wxFileName(BuiltInModulesDir(), wxT("resampler.pt"))
+                                       .GetFullPath());
+   try
+   {
+      mResampler = std::make_unique<torch::jit::script::Module>
+                     (torch::jit::load(resamplerPath, torch::kCPU));
+      mResampler->eval();
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error an error occurred while loading the resampler"), e.what());
+   }
+}
+
+void DeepModel::Load(const std::string &modelPath)
+{
+   try
+   { 
+      // set mResampler
+      LoadResampler();
+
+      // set mModel
+      mModel = std::make_unique<torch::jit::script::Module>
+                  (torch::jit::load(modelPath, torch::kCPU));
+      mModel->eval();
+      
+      // finally, mark as loaded
+      mLoaded = true;
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
```

Comment:
If you follow my first suggestion, `Cleanup()` would be eliminated (here and in all other places), and mutation of members of `*this` would happen only on the success paths and only after all possibly throwing steps are done.

Reply:
7ae287f72127a0f04e62564fc227c7c958ee76c1